plugins/prometheus_plugin/prometheus_plugin.cpp

```diff
@@ -0,0 +1,235 @@
+               pp->register_metrics_listener(create_metrics_listener("prod"));
+            } else {
+               dlog("producer_plugin not found -- metrics not added");
+            }
+
+            http_plugin* hp = app().find_plugin<http_plugin>();
+            if (nullptr != pp) {
+               _plugin_metrics.emplace(std::pair{"http", std::vector<runtime_metric>()});
+               hp->register_metrics_listener(create_metrics_listener("http"));
+            } else {
+               dlog("producer_plugin not found -- metrics not added");
+            }
+         }
+
+         std::string metrics() {
+            auto start_time_of_request = std::chrono::steady_clock::now();
+            metrics_model mm;
+            vector<runtime_metric> prometheus_metrics = _metrics.metrics();
+            mm.add_runtime_metrics(prometheus_metrics);
+
+            for (auto& pm : _plugin_metrics) {
+               mm.add_runtime_metrics(pm.second);
+            }
+
+            std::string body = mm.serialize();
+
+            auto stop_time_of_request = std::chrono::steady_clock::now();
+            auto duration = std::chrono::duration_cast<std::chrono::microseconds>(
+                  stop_time_of_request - start_time_of_request);
+
+            _metrics.bytes_transferred.value += body.length();
+            _metrics.num_scrapes.value++;
+
+            return body;
+         }
+
+         void metrics_async(chain::plugin_interface::next_function<std::string> results) {
+            _prometheus_strand.post([self=this, results=std::move(results)]() {
+               results(self->metrics());
+            });
+         }
+
+
+   };
+
+   using metrics_params = fc::variant_object;
+
+   struct prometheus_api {
+      fc::microseconds max_response_time_us{30*1000};
+      prometheus_plugin_impl& _pp;
+
+      fc::time_point start() const {
+         return fc::time_point::now() + max_response_time_us;
+      }
+
+      void metrics(const metrics_params& p, chain::plugin_interface::next_function<std::string> results) {
+         _pp.metrics_async(results);
+      }
+
+      prometheus_api(prometheus_plugin_impl& plugin) : _pp(plugin){
+      }
+   };
+
+   prometheus_plugin::prometheus_plugin() {
+      my.reset(new prometheus_plugin_impl{});
+      prometheus_api handle(*my);
+      app().get_plugin<http_plugin>().add_async_api({
+        CALL_ASYNC_WITH_400(prometheus, handle, eosio, metrics, std::string, 200, http_params_types::no_params)}, http_content_type::plaintext);
+   }
+
+   prometheus_plugin::~prometheus_plugin() {}
+
+   std::string prometheus_plugin::metrics() {return my->metrics();}
+
+   void prometheus_plugin::set_program_options(options_description&, options_description& cfg) {
+
+   }
+
+   void prometheus_plugin::plugin_initialize(const variables_map& options) {
+      my->initialize_metrics();
+   }
+
+   void prometheus_plugin::plugin_startup() {
+      my->_prometheus_thread_pool.start(1, []( const fc::exception& e ) {} );
```

Comment:
Log the exception. In this particular case we likely do not want to `app().quit()`, but open to other opinions.